livecheck: modified urls cop

nandahkrishna opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Changes to the urls cop, related to the migration of Livecheckables to `livecheck` blocks in Homebrew/homebrew-core Formulae.

Summary:
* `archey-osx`: Shortened the whitelist prefix
* `apache` urls - skipped livecheck urls. Modifying the url to the suggested alternative did not work with livecheck, which currently uses `archive.apache.org` urls for the version heuristic.
* `sourceforge` urls - skipped livecheck urls. **Question:** The alternative suggested is `download.sourceforge.net/...`, however that redirects to just `sourceforge.net/...`; and while I've just disabled the check specifically for livecheck urls, should it be removed as there's a redirection anyway?

Other related changes: PRs Homebrew/homebrew-livecheck#748, Homebrew/homebrew-livecheck#749, Homebrew/homebrew-livecheck#750, Homebrew/homebrew-livecheck#751, Homebrew/homebrew-livecheck#752

MikeMcQuaid commented on Library/Homebrew/rubocops/urls.rb:
```diff
@@ -46,8 +46,13 @@ class Urls < FormulaCop
           mirrors = find_every_func_call_by_name(body_node, :mirror)
 
+          # Identify livecheck urls, to disable some checks for them
+          livecheck = find_every_func_call_by_name(body_node, :livecheck).first
+          non_livecheck_urls -= find_every_func_call_by_name(livecheck.parent, :url) unless livecheck.nil?
```
```suggestion
          if (livecheck = find_every_func_call_by_name(body_node, :livecheck).first)
            non_livecheck_urls -= find_every_func_call_by_name(livecheck.parent, :url)
          end
```

nandahkrishna commented:
> Is it redirecting from HTTP to HTTPS or vice-versa? We have a few HTTPS-to-HTTP redirects because we hope in future they will result in/redirect to a fully HTTPS URL.

Ah that isn't the case. When you visit, say, https://downloads.sourceforge.net/project/offo/files/ it redirects you to https://sourceforge.net/projects/offo/files/ (probably due to some updates to SourceForge itself?).

Livecheck already uses only the `https://sourceforge.net/...` version, but `brew style` raises a style issue when it spots this kind of url, suggesting `https://downloads.sourceforge.net/...` instead.

While changing the urls to `downloads.sourceforge.net/...` _probably_ wouldn't cause an issue (I tested this with a few Livecheckables), there are too many Livecheckables to change. Furthermore, I'm doubtful if we should proceed with the change, due to the point I've mentioned in the first paragraph.

> Is this an issue with livecheck handling redirections or does the redirected-to URL not work?

No issues with livecheck when using either of the urls, at least for some formulae.

MikeMcQuaid commented:
> Ah that isn't the case. When you visit, say, https://downloads.sourceforge.net/project/offo/files/ it redirects you to https://sourceforge.net/projects/offo/files/ (probably due to some updates to SourceForge itself?).

I think this may vary from project to project 😬. 

> No issues with livecheck when using either of the urls, at least for some formulae.

I think ideally we'd use the format `brew style` suggests for all livecheckables and keep the check to be more consistent and avoid confusion. You could even investigate making an autocorrect rule for this so `brew style --fix` can do the work for you 😁.

That said, I don't feel strongly enough about this for the PR to be blocked on it so: up to you.

nandahkrishna commented:
I think skipping the SourceForge check for the livecheck urls (at least for now) might be a better option. @samford happened to mention that they tested all Livecheckables and ensured that all except 2 were working fine, so I don't really want to tamper with the urls.

If, however, we decide to go with the format suggested by `brew style`, it's just a small PR to make the changes 😅.

What do you think, @MikeMcQuaid?

samford commented:
The reason why these URLs continue to work in livecheckables is because livecheck isn't actually checking that particular URL. Instead, the URL in the livecheckable triggers the SourceForge strategy, which only uses the URL to derive the project name (e.g., `aa-project`) and formulate the RSS feed URL (e.g., https://sourceforge.net/projects/aa-project/rss), which it then checks.

Outside of the SourceForge strategy, the modified URLs typically redirect in the browser (e.g., https://downloads.sourceforge.net/project/asio/ goes to https://sourceforge.net/projects/asio/files/) but they 404 elsewhere (e.g., with Ruby or `curl`).

In general, I think we should only modify the `livecheck` block's URL when it's actually appropriate/beneficial. Even though the heuristic is smart enough to work with these bad https://downloads.sourceforge.net URLs, it doesn't make sense to use inappropriate URLs in the `livecheck` block just to appease a particular audit in `brew style` that doesn't apply in the context of livecheck.

The fundamental issue is that some of the existing `urls` audits assume that `url`s in the formula are for some sort of thing to download (an archive, a patch, etc.). This may be true currently but it won't be the case once `livecheck` blocks are added. URLs in livecheck are HTML pages, RSS feeds, Git repos, etc., so previous assumptions may not apply in the same way.

In my eyes, if it's not possible/appropriate to modify existing `urls` cops to avoid these kinds of assumptions, it makes sense to skip inappropriate checks for the `livecheck` URL.

MikeMcQuaid commented:
> Even though the heuristic is smart enough to work with these bad https://downloads.sourceforge.net URLs, it doesn't make sense to use inappropriate URLs in the `livecheck` block just to appease a particular audit in `brew style` that doesn't apply in the context of livecheck.

I'm not sure what makes it inappropriate if it's just used for a heuristic? The lack of consistency here will confuse some users and the lack of checks for the value will result in incorrect results.

> URLs in livecheck are HTML pages, RSS feeds, Git repos, etc., so previous assumptions may not apply in the same way.

If these URLs can be inferred from the download URL: this seems preferable to having to duplicate them with a non-audited one?

> The current implementation accomplishes the goal but it does so in a way that only works for the livecheck URL. That is to say, what happens if we need to skip another URL in the future?

I'm not sure I understand this? We already have some whitelisting constants for other things but we don't want to be whitelisting arbitrary URLs in homebrew-core.

samford commented:
> I'm not sure what makes it inappropriate if it's just used for a heuristic?

Let's put aside the behavior of the heuristic and SourceForge strategy for a moment.

The intention of the `Use https://downloads.sourceforge.net to get geolocation` check is to ensure sourceforge.net download `url`s use https://downloads.sourceforge.net, as the download URL will redirect to a mirror (ideally close to the user).

This check was written with the assumption that any sourceforge.net `url` in the formula would be for some sort of download and that will no longer be true after `livecheck` blocks are added. A sourceforge.net `url` in a `livecheck` block is an HTML page or RSS feed, so it's not appropriate to execute this particular check in this context.

downloads.sourceforge.net shouldn't be used for these livecheck URLs and SourceForge will correctly redirect you to sourceforge.net (but seemingly only in browsers, as these URLs 404 in Ruby). This redirection is SourceForge saying, "downloads.sourceforge.net is meants for downloads. This is an HTML page/RSS feed, so you should be using sourceforge.net".

---

> The lack of consistency here will confuse some users and the lack of checks for the value will result in incorrect results.

I'm not advocating for skipping all the `urls` checks for a livecheck URL but rather just the ones that don't apply due to the difference in context, as in this PR. We're already skipping specific checks when URLs meet a given condition (e.g., `next if url.include? "/svnroot/"`) and this should be similar.

---

> If these URLs can be inferred from the download URL: this seems preferable to having to duplicate them with a non-audited one?

When a livecheckable is using a SourceForge URL as a way of enforcing use of the SourceForge strategy and using a preexisting URL in the formula would suffice, I agree. With additional livecheck features, we could do something like `url :stable` and/or `using :sourceforge` to avoid duplicating the URL.

However, once the `using` feature is implemented, I'll be adding 7+ livecheckables that use a SourceForge URL and the `page_match` strategy. There are times where a formula's software is old enough that it no longer appears in the SourceForge RSS feed (having been pushed out by newer releases of different software in the same project) and we have to check the files/folders on a SourceForge page instead. Modifying the URL to use downloads.sourceforge.net to appease `brew style` would break the check in this case, due to the resulting URL being a 404 in Ruby instead of a redirection.

---

> I'm not sure I understand this? We already have some whitelisting constants for other things but we don't want to be whitelisting arbitrary URLs in homebrew-core.

What I'm trying to say is that this particular setup isn't easily extended in the future. For example, say we added a `foo` block with a `url` that's also not a download and should skip certain `urls` checks. Assumedly we would have to replicate the existing setup to create a `non_foo_urls` array.

But what happens when we need to skip a check for both `livecheck` and `foo`? Do we then create a `non_foo_livecheck_urls` array? What happens if we add `bar` and need to skip all three, `non_bar_foo_livecheck_urls`?

Creating these arrays once _does_ avoid unnecessary allocations (compared to subtracting URLs from `urls` on the fly for specific audits) but it's not necessary. We could simply store the livecheck URL itself (the URL string, not the node) and use that in the checks we want to skip (e.g., `next if url == livecheck_url`, `if ... && url != livecheck_url`, etc.). I've tested this locally and it works as expected.

With this alternative approach, we could easily skip checks for additional URLs without having to create arrays for each. Similarly, this allows us to skip specific checks that are lumped together into one `audit_urls` call (e.g., the SourceForge checks), instead of having to skip them all.

---

Let me know if any of this is unclear, as we may be misunderstanding each other.

MikeMcQuaid commented:
> Modifying the URL to use downloads.sourceforge.net to appease `brew style` would break the check in this case, due to the resulting URL being a 404 in Ruby instead of a redirection.

That's a good enough reason for me to skip the sourceforge case 👍 



> For example, say we added a `foo` block with a `url` that's also not a download and should skip certain `urls` checks.

https://en.wikipedia.org/wiki/You_aren't_gonna_need_it 😉. We can handle that situation if/when it comes up.



> With this alternative approach, we could easily skip checks for additional `url`s without having to create arrays for each.

Just so I'm on the same page here: is this primarily a performance concern? If so, what are the benchmarks before/after this change?

Thanks!

samford commented:
> Just so I'm on the same page here: is this primarily a performance concern?

Not in my eyes. There isn't going to be a really noticeable benchmark/memory difference between the setup in this PR and the proposed alternative, even if you're running `brew style` across all the formulae.

This is about allowing for skipping at the level of a specific check (instead of all checks in a given `audit_urls` block) and bringing the implementation in line with how skipping is currently handled in the `urls` checks (i.e., a conditional doing some comparison of the `url`). Coincidentally, it's also easy to extend in the future if an additional `url` is added to the formula that needs to be skipped.

I don't see any compelling reasons to use the `non_livecheck_urls` setup when there are clear benefits to the aforementioned alternative. Does this make sense?

MikeMcQuaid commented:
> I don't see any compelling reasons to use the `non_livecheck_urls` setup when there are clear benefits to the aforementioned alternative. Does this make sense?

So: specifically whitelist the prefixes we want to allow only for livechecks? Won't the prefixes in those cases be more-or-less exactly what the match is? I'm not sure I understand what you're proposing.

Just to be explicit: do you consider this PR blocked on resolving this or not? I think it's fine to go in as-is, personally.

samford commented:
Let me try to put this into context, as that may help to explain it.

The implementation in this PR works by creating a separate array of the `urls` in the formula that doesn't include the livecheck `url`: `non_livecheck_urls`. This is then used in `audit_urls` calls in place of `urls`.

Some `audit_urls` calls only have one check in the block:

```ruby
apache_pattern = %r{^https?://(?:[^/]*\.)?apache\.org/(?:dyn/closer\.cgi\?path=/?|dist/)(.*)}i
audit_urls(non_livecheck_urls, apache_pattern) do |match, url|
  problem "#{url} should be `https://www.apache.org/dyn/closer.lua?path=#{match[1]}`"
end
```

The checks for SourceForge URLs are all part of one big block:

```ruby
sourceforge_patterns = %r{^https?://.*\b(sourceforge|sf)\.(com|net)}
audit_urls(non_livecheck_urls, sourceforge_patterns) do |_, url|
  # Skip if the URL looks like a SVN repo
  next if url.include? "/svnroot/"
  next if url.include? "svn.sourceforge"
  next if url.include? "/p/"

  if url =~ /(\?|&)use_mirror=/
    problem "Don't use #{Regexp.last_match(1)}use_mirror in SourceForge urls (url is #{url})."
  end

  problem "Don't use /download in SourceForge urls (url is #{url})." if url.end_with?("/download")

  if url.match?(%r{^https?://sourceforge\.})
    problem "Use https://downloads.sourceforge.net to get geolocation (url is #{url})."
  end

  if url.match?(%r{^https?://prdownloads\.})
    problem <<~EOS.chomp
      Don't use prdownloads in SourceForge urls (url is #{url}).
              See: http://librelist.com/browser/homebrew/2011/1/12/prdownloads-is-bad/
    EOS
  end

  if url.match?(%r{^http://\w+\.dl\.})
    problem "Don't use specific dl mirrors in SourceForge urls (url is #{url})."
  end

  problem "Please use https:// for #{url}" if url.start_with? "http://downloads"
end
```

With this PR, all of the SourceForge checks are skipped for the `livecheck` URL. The others aren't yet offensive for the existing SourceForge URL, so we only _need_ to skip the `downloads.sourceforge.net` check but we can't do that with this PR's implementation.

I can send you a patch on Slack, so you can see what the alternative implementation could look like and how it differs from what's in this PR.

> Just to be explicit: do you consider this PR blocked on resolving this or not? I think it's fine to go in as-is, personally.

This is just my view but I don't see any reason to merge this as-is when it has shortcomings and there's a better alternative that's just as easy to implement.

nandahkrishna commented:
I'll work on updating this to what you've suggested @samford.

MikeMcQuaid commented:
> With this PR, all of the SourceForge checks are skipped for the `livecheck` URL. The others aren't yet offensive for the existing SourceForge URLs in livecheckables, so we only _need_ to skip the `downloads.sourceforge.net` check but we can't do that with this PR's implementation.

> I sent you a patch on Slack, so you can see what the alternative implementation could look like and how it differs from what's in this PR.

Ah, I see now from your suggested patch and I agree 👍. Sorry for my misunderstanding before!

Essentially: get a single `livecheck_url` and check that against a given `url` when iterating through them rather than pretending they don't exist. This seems even better for me given what I was wanting so: thanks again @samford ❤️ !

nandahkrishna commented:
I've modified my implementation based on @samford's suggestions. In the process, some other changes as well:
* Added https://artifacts.videolan.org/x264/release-macos/ to the whitelist as well, which was incorrectly checked and reported as a binary package's url.
* Skipped the `maven` check for `livecheck` urls, because we use the `https://repo1.maven.org/...` url for one Livecheckable (`closure-compiler`) and that caused a style issue. However the url suggested by the cop works in livecheck as well, so I could create a PR there and not skip the `maven` check here.

nandahkrishna commented:
I've created Homebrew/homebrew-livecheck#762 to fix the url issue for `closure-compiler`.

nandahkrishna commented:
> I don't think we need the # This check doesn't apply to livecheck urls comments

Okay I'll remove them 👍🏼

> We probably don't want to skip the Maven check.

Yeah my latest commit changed it back to not skipping the Maven check. I've created a PR in homebrew-livecheck to change the url (I've mentioned it in a previous comment).

samford commented:
I don't have any specific issues with the current version of this PR and I'll defer to @MikeMcQuaid's review at this point. Thanks for the rework, Nanda.

MikeMcQuaid commented:
Thanks again @nandahkrishna!
